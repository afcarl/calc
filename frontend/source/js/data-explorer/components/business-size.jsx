import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { filterActive } from '../util';
import { makeOptions } from './util';
import { setBusinessSize as setBusinessSizeAction } from '../actions';
import { BUSINESS_SIZE_LABELS } from '../constants';

export function BusinessSize({ idPrefix, size, setSize }) {
  const id = `${idPrefix}business_size`;
  const handleChange = (e) => { setSize(e.target.value); };

  return (
    <div className="filter filter-business_size">
      <label htmlFor={id}>
Business size:
      </label>
      <select
        id={id}
        name="business_size"
        value={size}
        onChange={handleChange}
        className={filterActive(size !== '')}
      >
        {makeOptions(BUSINESS_SIZE_LABELS)}
      </select>
    </div>
  );
}

BusinessSize.propTypes = {
  size: PropTypes.string.isRequired,
  setSize: PropTypes.func.isRequired,
  idPrefix: PropTypes.string,
};

BusinessSize.defaultProps = {
  idPrefix: '',
};

export default connect(
  state => ({ size: state.business_size }),
  { setSize: setBusinessSizeAction },
)(BusinessSize);
