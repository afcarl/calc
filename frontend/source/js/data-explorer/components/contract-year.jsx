import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import Tooltip from './tooltip';
import { setContractYear as setContractYearAction } from '../actions';
import {
  CONTRACT_YEAR_CURRENT,
  CONTRACT_YEAR_1,
  CONTRACT_YEAR_2,
  CONTRACT_YEAR_LABELS,
} from '../constants';

const YEAR_LI_INFO = {
  [CONTRACT_YEAR_CURRENT]: {
    className: 'contract-current-year-span checkbox-focus',
    shortLabel: 'Current',
    idSuffix: 'current-year',
  },
  [CONTRACT_YEAR_1]: {
    className: 'checkbox-focus',
    shortLabel: '+1',
    idSuffix: 'one-year-out',
  },
  [CONTRACT_YEAR_2]: {
    className: 'contract-last-year-span checkbox-focus',
    shortLabel: '+2',
    idSuffix: 'two-years-out',
  },
};

const TOOLTIP = 'Show rates for the initial year of the contract, or for subsequent years (+1 or +2).';

export function ContractYear({ idPrefix, contractYear, setContractYear }) {
  const listItem = (year) => {
    const { className, shortLabel, idSuffix } = YEAR_LI_INFO[year];
    const id = `${idPrefix}${idSuffix}`;

    return (
      <li className="contract-list-item">
        <label htmlFor={id} className="radio">
          <input
            id={id}
            type="radio"
            checked={contractYear === year}
            onChange={() => { setContractYear(year); }}
            name="contract-year"
            value={year}
            tabIndex="0"
          />
          <span
            tabIndex="-1"
            className={className}
            aria-hidden="true"
          >
            {shortLabel}
          </span>
          <span className="usa-sr-only">
            {CONTRACT_YEAR_LABELS[year]}
          </span>
        </label>
      </li>
    );
  };

  return (
    <div className="filter contract-year">
      <fieldset className="fieldset-inputs">

        <legend>
Contract year:
        </legend>

        <span className="filter-more-info">
          <Tooltip text={TOOLTIP}>
            <a /* eslint-disable-line jsx-a11y/anchor-is-valid */
              href=""
              aria-label={TOOLTIP}
              onClick={(e) => { e.preventDefault(); }}
            >
              What&apos;s this?
            </a>
          </Tooltip>
        </span>

        <h3 className="usa-sr-only">
Contract Year
        </h3>

        <ul className="contract-year-block">
          {listItem(CONTRACT_YEAR_CURRENT)}
          {listItem(CONTRACT_YEAR_1)}
          {listItem(CONTRACT_YEAR_2)}
        </ul>
      </fieldset>
    </div>
  );
}

ContractYear.propTypes = {
  contractYear: PropTypes.string.isRequired,
  setContractYear: PropTypes.func.isRequired,
  idPrefix: PropTypes.string,
};

ContractYear.defaultProps = {
  idPrefix: '',
};

export default connect(
  state => ({ contractYear: state['contract-year'] }),
  { setContractYear: setContractYearAction },
)(ContractYear);
