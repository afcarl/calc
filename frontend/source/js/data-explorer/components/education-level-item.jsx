import PropTypes from 'prop-types';
import React from 'react';

import { autobind } from '../util';
import { EDU_LABELS } from '../constants';

export default class EducationLevelItem extends React.Component {
  constructor(props) {
    super(props);
    autobind(this, ['_onClick']);
  }

  _onClick() {
    this.props.onCheckboxClick(this.props.value);
  }

  render() {
    return (
      <li>
        <input
          id={this.props.id}
          type="checkbox"
          value={this.props.value}
          checked={this.props.checked}
          onChange={this._onClick}
          name="education"
        />
        <label htmlFor={this.props.id}>
          {EDU_LABELS[this.props.value]}
        </label>
      </li>
    );
  }
}

EducationLevelItem.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  checked: PropTypes.bool.isRequired,
  onCheckboxClick: PropTypes.func.isRequired,
};
