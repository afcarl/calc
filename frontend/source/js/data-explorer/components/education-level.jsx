/* global document */

import PropTypes from 'prop-types';

import React from 'react';
import { connect } from 'react-redux';

import SlideyPanel from './slidey-panel';
import EducationLevelItem from './education-level-item';

import {
  autobind,
  handleEnterOrSpace,
  filterActive,
} from '../util';

import {
  EDU_LABELS,
} from '../constants';

import {
  toggleEducationLevel,
} from '../actions';

// TODO: We could just use jQuery for this, but I wanted to decouple
// the new React code from jQuery as much as possible for now.
function elementContains(container, contained) {
  let target = contained.parentNode;

  while (target) {
    if (target === container) {
      return true;
    }
    target = target.parentNode;
  }
  return false;
}

/**
 * The following logic was created to mimic the following
 * legacy jQuery behavior:
 *
 *   Dropdown with Multiple checkbox select with jQuery - May 27, 2013
 *   (c) 2013 @ElmahdiMahmoud
 *   license: http://www.opensource.org/licenses/mit-license.php
 */

export class EducationLevel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
    };
    autobind(this, ['handleToggleMenu', 'handleDocumentClick',
      'handleCheckboxClick']);
  }

  componentDidMount() {
    document.addEventListener('click', this.handleDocumentClick);
    document.addEventListener('focus', this.handleDocumentClick, true);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleDocumentClick);
    document.removeEventListener('focus', this.handleDocumentClick, true);
  }

  handleDocumentClick(e) {
    if (!this.state.expanded) {
      return;
    }

    // Whenever users click outside of our dropdown, hide it.
    if (!elementContains(this.dropdownEl, e.target)) {
      this.setState({ expanded: false });
    }
  }

  handleToggleMenu(e) {
    // TODO: The original jQuery logic used .slideDown() here, it'd
    // be nice if we could use some sort of transition too.

    e.preventDefault();
    this.setState({
      expanded: !this.state.expanded, /* eslint-disable-line react/no-access-state-in-setstate */
    });
  }

  handleCheckboxClick(level) {
    this.props.toggleEducationLevel(level);
  }

  render() {
    const { levels, idPrefix } = this.props;
    const inputs = Object.keys(EDU_LABELS).map((value) => {
      const id = idPrefix + value;
      return (
        <EducationLevelItem
          key={value}
          id={id}
          checked={levels.indexOf(value) >= 0}
          value={value}
          onCheckboxClick={this.handleCheckboxClick}
        />
      );
    });
    let linkContent;

    if (levels.length === 0) {
      linkContent = (
        <span className="eduSelect">
Select
          <span className="usa-sr-only">
            {' '}
to reveal Education Level options
          </span>
        </span>
      );
    } else {
      const selectedLevels = levels.map((value) => {
        const label = EDU_LABELS[value];
        return (
          <span key={value} title={label}>
            {label}
          </span>
        );
      });
      linkContent = (
        <div className="multiSel">
          {selectedLevels}
        </div>
      );
    }

    const eduLevelId = `${this.props.idPrefix}education_level`;

    return (
      <div>
        <label htmlFor={eduLevelId}>
Education level:
        </label>
        <dl
          id={eduLevelId}
          className="dropdown"
          ref={(el) => { this.dropdownEl = el; }}
        >
          <dt>
            <a /* eslint-disable-line jsx-a11y/anchor-is-valid */
              href=""
              onClick={this.handleToggleMenu}
              role="button"
              aria-expanded={this.state.expanded.toString()}
              onKeyDown={handleEnterOrSpace(this.handleToggleMenu)}
              className={filterActive(levels.length !== 0)}
            >
              {linkContent}
            </a>
          </dt>

          <dd>
            <div className="multiSelect">
              <fieldset>
                <legend className="usa-sr-only">
Education level:
                </legend>

                <SlideyPanel
                  component="ul"
                  expanded={this.state.expanded}
                >
                  {inputs}
                </SlideyPanel>
              </fieldset>
            </div>
          </dd>
        </dl>
      </div>
    );
  }
}

EducationLevel.propTypes = {
  levels: PropTypes.array.isRequired,
  idPrefix: PropTypes.string,
  toggleEducationLevel: PropTypes.func.isRequired,
};

EducationLevel.defaultProps = {
  idPrefix: 'education-level-',
};

export default connect(
  state => ({ levels: state.education }),
  { toggleEducationLevel },
)(EducationLevel);
