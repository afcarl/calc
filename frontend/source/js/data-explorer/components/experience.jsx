/* eslint-disable no-unused-vars */

import React from 'react';
import { connect } from 'react-redux';

/* global $ */

import PropTypes from 'prop-types';

import { Range } from 'rc-slider';

import {
  setExperience,
} from '../actions';

import {
  autobind,
  filterActive,
} from '../util';

import {
  MIN_EXPERIENCE,
  MAX_EXPERIENCE,
} from '../constants';

function makeOptions(min, max) {
  const options = [];

  for (let i = min; i <= max; i++) {
    options.push(
      <option key={i} value={i}>
        {i}
      </option>
    );
  }

  return options;
}

export class Experience extends React.Component {
  constructor(props) {
    super(props);
    autobind(this, ['onSliderChange', 'onAfterSliderChange']);
    this.state = {
      sliderVal: [this.props.min, this.props.max],
    };
  }

  componentWillReceiveProps(newProps) {
    const [min, max] = this.state.sliderVal;
    if (min !== newProps.min || max !== newProps.max) {
      this.setState({ sliderVal: [newProps.min, newProps.max] });
    }
  }

  onSliderChange(val) {
    this.setState({ sliderVal: val });
  }

  onAfterSliderChange(val) {
    const [min, max] = val;
    if (min !== this.props.min) {
      this.props.setExperience('min', min);
    }
    if (max !== this.props.max) {
      this.props.setExperience('max', max);
    }
  }

  render() {
    const [min, max] = this.state.sliderVal;

    const rangeId = `${this.props.idPrefix}range`;
    const minId = `${this.props.idPrefix}min`;
    const maxId = `${this.props.idPrefix}max`;
    const baseClasses = 'select-small';
    const minClasses = filterActive(min !== MIN_EXPERIENCE, baseClasses);
    const maxClasses = filterActive(max !== MAX_EXPERIENCE, baseClasses);

    const onChange = type => (e) => {
      this.props.setExperience(type, parseInt(e.target.value, 10));
    };

    return (
      <div className="filter" ref={(el) => { this.rootEl = el; }}>
        <fieldset>
          <legend>
            Experience:
          </legend>
          <Range
            id={rangeId}
            allowCross={false}
            min={MIN_EXPERIENCE}
            max={MAX_EXPERIENCE}
            value={this.state.sliderVal}
            onChange={this.onSliderChange}
            onAfterChange={this.onAfterSliderChange}
            className="experience-slider"
          />
          <div className="experience_range">
            <label htmlFor={minId} className="usa-sr-only">
Minimum Years
            </label>
            <select
              id={minId}
              value={min}
              name="min_experience"
              onChange={onChange('min')}
              className={minClasses}
            >
              {makeOptions(MIN_EXPERIENCE, max)}
            </select>
            {' - '}
            <label htmlFor={maxId} className="usa-sr-only">
Maximum Years
            </label>
            <select
              id={maxId}
              value={max}
              name="max_experience"
              onChange={onChange('max')}
              className={maxClasses}
            >
              {makeOptions(min, MAX_EXPERIENCE)}
            </select>
            {' years'}
          </div>
        </fieldset>
      </div>
    );
  }
}

Experience.propTypes = {
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
  setExperience: PropTypes.func.isRequired,
  idPrefix: PropTypes.string,
};

Experience.defaultProps = {
  idPrefix: 'experience-',
};

export default connect(
  state => ({
    min: state.min_experience,
    max: state.max_experience,
  }),
  { setExperience },
)(Experience);
