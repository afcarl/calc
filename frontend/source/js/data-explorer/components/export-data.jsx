import PropTypes from 'prop-types';
import * as qs from 'querystring';

import React from 'react';
import { connect } from 'react-redux';

import { getRatesParameters } from '../rates-request';
import { API_BASE, API_PATH_RATES_CSV } from '../api';

export function ExportData({ querystring }) {
  const href = `${API_BASE}${API_PATH_RATES_CSV}/${querystring}`;

  return (
    <a
      className="usa-button usa-button-primary export-data"
      title="Click to export your search results to an Excel file (CSV)"
      href={href}
    >
â¬‡ Export data (CSV)
    </a>
  );
}

ExportData.propTypes = {
  querystring: PropTypes.string.isRequired,
};

function mapStateToProps(state) {
  return {
    querystring: `?${qs.stringify(getRatesParameters(state))}`,
  };
}

export default connect(mapStateToProps)(ExportData);
