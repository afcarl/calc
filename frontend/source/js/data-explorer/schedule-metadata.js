// @ts-check

/** @typedef {import('./api').ScheduleMetadata} ScheduleMetadata */

/**
 * Mapping from schedule identifiers (e.g. "MOBIS") to
 * schedule labels (e.g. "874 - Legacy MOBIS").
 */
export const scheduleLabels = {};

/**
 * Populate scheduleLabels from schedule metadata.
 * 
 * @param {ScheduleMetadata[]} schedules 
 */
export function populateScheduleLabels(schedules) {
  Object.keys(scheduleLabels).forEach((label) => {
    delete scheduleLabels[label];
  });
  schedules.forEach((schedule) => {
    scheduleLabels[schedule.schedule] = schedule.full_name;
  });
}
