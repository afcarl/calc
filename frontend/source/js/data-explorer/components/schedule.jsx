import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { filterActive } from '../util';
import { makeOptions } from './util';
import { setSchedule as setScheduleAction } from '../actions';
import { scheduleLabels } from '../schedule-metadata';

export function Schedule({ idPrefix, schedule, setSchedule }) {
  const id = `${idPrefix}schedule`;
  const handleChange = (e) => { setSchedule(e.target.value); };

  return (
    <div className="filter filter-schedule">
      <label htmlFor={id}>
Contract vehicle:
      </label>
      <a href="/about#schedules" className="filter-more-info">
        What&apos;s this?
      </a>
      <select
        id={id}
        name="schedule"
        value={schedule}
        onChange={handleChange}
        className={filterActive(schedule !== '')}
      >
        {makeOptions(scheduleLabels)}
      </select>
    </div>
  );
}

Schedule.propTypes = {
  schedule: PropTypes.string.isRequired,
  setSchedule: PropTypes.func.isRequired,
  idPrefix: PropTypes.string,
};

Schedule.defaultProps = {
  idPrefix: '',
};

export default connect(
  state => ({ schedule: state.schedule }),
  { setSchedule: setScheduleAction },
)(Schedule);
