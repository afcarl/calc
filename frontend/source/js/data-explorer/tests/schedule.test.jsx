import toJson from 'enzyme-to-json';

import { Schedule } from '../components/schedule';
import makeSetup from './testSetup';
import { scheduleLabels } from '../schedule-metadata';

const defaultProps = {
  schedule: '',
  idPrefix: 'zzz_',
  setSchedule: jest.fn(),
};

const setup = makeSetup(Schedule, defaultProps);

describe('<Schedule>', () => {
  it('renders correctly', () => {
    const { wrapper } = setup();
    const select = wrapper.find('select#zzz_schedule');
    expect(select.exists()).toBeTruthy();
    const options = wrapper.find('option');
    // should have options for all schedules + 1 for "(all)"
    expect(options.length).toBe(Object.keys(scheduleLabels).length + 1);
    Object.keys(scheduleLabels).forEach((sched) => {
      const title = scheduleLabels[sched];
      const opt = wrapper.find(`option[value="${sched}"]`);
      expect(opt.exists()).toBeTruthy();
      expect(opt.text()).toBe(title);
    });
  });

  it('matches snapshot', () => {
    const { wrapper } = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('calls setSchedule on change', () => {
    const { props, mounted } = setup();
    expect(props.setSchedule.mock.calls.length).toBe(0);

    mounted.find('select#zzz_schedule')
      .simulate('change', { target: { value: 'MOBIS' } });

    expect(props.setSchedule.mock.calls.length).toBe(1);
    expect(props.setSchedule.mock.calls[0][0]).toBe('MOBIS');
  });

  it('has filter_active class when levels are selected', () => {
    let { wrapper } = setup({ schedule: '' });
    expect(wrapper.find('select.filter_active').exists()).toBeFalsy();

    ({ wrapper } = setup({ schedule: 'PES' }));
    expect(wrapper.find('select.filter_active').exists()).toBeTruthy();
  });
});
