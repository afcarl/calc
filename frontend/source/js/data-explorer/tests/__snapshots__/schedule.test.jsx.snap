// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Schedule> matches snapshot 1`] = `
<div
  className="filter filter-schedule"
>
  <label
    htmlFor="zzz_schedule"
  >
    Contract vehicle:
  </label>
  <a
    className="filter-more-info"
    href="/about#schedules"
  >
    What's this?
  </a>
  <select
    className=""
    id="zzz_schedule"
    name="schedule"
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      (all)
    </option>
    <option
      key="Environmental"
      value="Environmental"
    >
      899 - Legacy Environmental
    </option>
    <option
      key="Logistics"
      value="Logistics"
    >
      87405 - Legacy Logistics
    </option>
    <option
      key="MOBIS"
      value="MOBIS"
    >
      874 - Legacy MOBIS
    </option>
    <option
      key="PES"
      value="PES"
    >
      871 - Legacy PES
    </option>
    <option
      key="Language Services"
      value="Language Services"
    >
      73802 - Legacy Language
    </option>
    <option
      key="AIMS"
      value="AIMS"
    >
      541 - Legacy AIMS
    </option>
    <option
      key="FABS"
      value="FABS"
    >
      520 - Legacy FABS
    </option>
    <option
      key="IT Schedule 70"
      value="IT Schedule 70"
    >
      132 - IT 70
    </option>
    <option
      key="Consolidated"
      value="Consolidated"
    >
      Consolidated
    </option>
  </select>
</div>
`;
